### PR TITLE
8332528: Generate code in SwitchBootstraps.generateTypeSwitch that require fewer adaptations

### PR DESCRIPTION
**Pull request by cl4es**
We can fold the call to `Objects.checkIndex` into the code generated in generateTypeSwitchSkeleton instead of doing so by filtering the MH argument. This loads 9 less classes (of which 8 generated LFs and Species classes) on a minimal test, while being neutral on a throughput sanity test:

```
Name                   Cnt  Base   Error   Test   Error  Unit  Change
SwitchSanity.switchSum  15 8,162 ± 0,117  8,152 ± 0,131 ns/op   1,00x (p = 0,800 )
  * = significant
  ```

A few additional optimizations includes moving some seldom used `findStatic` calls to a holder. All in all this means a reduction by 22M cycles to bootstrap a trivial switch expression on my M1.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332528](https://bugs.openjdk.org/browse/JDK-8332528): Generate code in SwitchBootstraps.generateTypeSwitch that require fewer adaptations (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author) ⚠️ Review applies to [f04d78ea](https://git.openjdk.org/jdk/pull/19307/files/f04d78ea53ed0074026311f82eb0d4eafee3438d)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19307/head:pull/19307` \
`$ git checkout pull/19307`

Update a local copy of the PR: \
`$ git checkout pull/19307` \
`$ git pull https://git.openjdk.org/jdk.git pull/19307/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19307`

View PR using the GUI difftool: \
`$ git pr show -t 19307`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19307.diff">https://git.openjdk.org/jdk/pull/19307.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19307#issuecomment-2121191868)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 1]
```diff
@@ -407,6 +400,11 @@ private static Consumer<CodeBuilder> generateTypeSwitchSkeleton(Class<?> selecto
         int EXTRA_CLASS_LABELS  = 3;
 
         return cb -> {
+            // Objects.checkIndex(RESTART_IDX, labelConstants + 1)
+            cb.iload(RESTART_IDX);
+            cb.loadConstant(labelConstants.length + 1);
+            cb.invokestatic(CD_Objects, "checkIndex", MethodTypeDesc.of(ConstantDescs.CD_int, ConstantDescs.CD_int, ConstantDescs.CD_int));
```
Comment (liach) [inline src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:406]: We should cache this MethodTypeDesc too.

Comment (cl4es) [reply src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:406]: `MethodTypeDesc.of` is actually quite cheap when inputs are `ClassDesc`s. Besides I think the main focus in these bootstraps should be improving the overhead of the first few calls - reduce dependencies, reduce runtime code generation - not treat this code as something that will be run many times over in a hot loop.

**Comment by cl4es** [comment]
To help evaluate and diagnose the startup overheads I added a `main` method to the provided `SwitchSanity` micro. This only runs a single invocation and is easy to run using your startup benchmarking script/tool of choice:

```
Name             Cnt          Base         Error           Test         Error         Unit  Change
Perfstartup-JMH   20        75,000 ±       4,455         62,000 ±       6,042        ms/op   1,21x (p = 0,000*)
  :.cycles           277393501,850 ± 4422757,249  253819501,350 ± 5482340,325       cycles   0,92x (p = 0,000*)
  :.instructions     701618995,500 ± 3774721,992  622929549,250 ± 3634060,803 instructions   0,89x (p = 0,000*)
  :.taskclock               82,000 ±       3,564         74,000 ±       4,365           ms   0,90x (p = 0,000*)
  * = significant
```

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 2]
```diff
@@ -621,11 +622,15 @@ private static MethodHandle generateTypeSwitch(MethodHandles.Lookup caller, Clas
...
         List<Class<?>> extraClassLabels = new ArrayList<>();
 
-        byte[] classBytes = ClassFile.of().build(ClassDesc.of(typeSwitchClassName(caller.lookupClass())),
+        byte[] classBytes = ClassFile.of().build(ReferenceClassDescImpl.ofValidatedBinaryName(typeSwitchClassName(caller.lookupClass())),
                 clb -> {
                     clb.withFlags(AccessFlag.FINAL, AccessFlag.SUPER, AccessFlag.SYNTHETIC)
                        .withMethodBody("typeSwitch",
-                                       TYPES_SWITCH_DESCRIPTOR,
+                                       MethodTypeDescImpl.ofValidated(ConstantDescs.CD_int,
+                                            ClassDesc.ofDescriptor(selectorType.descriptorString()),
```
Comment (liach) [inline src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:630]: `selectorType` can be primitive, therefore verification errors arise at `aload` in the generated code. This is why tests fail on GitHub actions.

**Review by liach — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java (hunk 2 above)
Comment (cl4es) [reply src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:630]: Dang. I had focused on making the `java/lang/runtime` tests pass locally, thinking that'd be sufficient. Luckily bf68b1d had only a minor benefit. Backing it out and tidying up a bit gives me these numbers:
```
Wall clock:    65.0 ms/op
.taskclock:    75.0 ms/op
.cycles:       255323158 average cycles elapsed
.instructions: 626086160 average instructions retired
```

**Comment by cl4es** [comment]
By adding a trivial type switch to `HelloClasslist` we archive a set of classes and avoid some runtime class generation. This grows the default CDS archive by approximately 1.2Mb by pulling in a large number of Classfile API classes, which is bound to happen soon anyway. 

On the SwitchSanity test, bootstrap cost drops by another 35M cycles:
```
Wall clock:    51.0 ms/op
.taskclock:    63.5 ms/op
.cycles:       220064872 average cycles elapsed
.instructions: 544981276 average instructions retired
```

**Review by liach — APPROVED** [review]

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java [hunk 3]
```diff
@@ -48,6 +49,9 @@
 import java.lang.classfile.ClassFile;
 import java.lang.classfile.Label;
 import java.lang.classfile.instruction.SwitchCase;
+
+import jdk.internal.constant.MethodTypeDescImpl;
```
Comment (lahodaj) [inline src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:53]: Nit - this import seems to be unused, and neither seem to be the changes to `MethodTypeDescImpl`. Is there some use missing? (OTOH, I like the `StaticHolders` for the stuff that relates to the enum-switch special case, for the time being.)

**Review by lahodaj — COMMENTED** [review]
Seems goo to me, with one nit related to `MethodTypeDescImpl`.

File: src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java (hunk 3 above)
Comment (cl4es) [reply src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java:53]: Oops, some leftovers from the optimizations I reverted. Fixed.

**Review by lahodaj — APPROVED** [review]
Looks good, thanks!

**Comment by cl4es** [comment]
Thanks!

/integrate

**Pull request closed without merging**